zkllvm/best-practices-limitations/using-stat-structures.mdx

```diff
@@ -21,18 +39,17 @@ Consider the following examples in which const size containers are used to build
     [[circuit]] bool merkle_tree_poseidon (
         typename pallas::base_field_type::value_type expected_root,
         [[private_input]] std::array<typename pallas::base_field_type::value_type, 0x20> layer_0_leaves) {
-
             std::array<typename pallas::base_field_type::value_type, 0x10> layer_1_leaves;
-            std::size_t layer_1_size = 0x10;
+            std::size_t layer_1_size = layer_1_leaves.size();
```

Comment:
I'm afraid it could be still not enough. I meant usage `.size()` directly in the loop statement instead of temporary variables. Please add `constexpr` modifier here, it's the best way to ensure that the loop can be unrolled.